### PR TITLE
refactor: improved options parsing in components (stage 1)

### LINKED ISSUE #1309: Refactor options parsing to use new GenericOptionsParser handling
More robust options parsing

### PR DESCRIPTION
**Pull request by msavy**
Idea is as follows:

* A robust and unified way of parsing `Map<String, String>` options that we have throughout the codebase which come from users
* Standardised error messages, if the implementor wants to use them
* Easily extended for custom options parsers
* Parser functions for all very common use-cases such as long, int, bool, String, path/file, etc.
* Allow predicates, with common set provided (e.g. greater than zero, any of, etc)
* Required and non-required inputs. 
  * Default values where appropriate

Start converting over one of the Elasticsearch modules, and then over time we can convert these over.

The rationale for this is that:
* Different areas of code handle user input differently, showing a wide variety of different errors
* Different null handling in different places
* Different invalid input handling
  * Sometimes it's an error, other places it causes 'default value', other times it is just accepted and breaks later.
* Reduce boilerplate and increase robustness of parsing code.
* Exceptions that contain useful information about what went wrong!

---

TL;DR: 

- Use `GenericOptionsParser` directly or extend it. 
- Instantiate via `new GenericOptionsParser(config)` 
- Pull values out with the various `get` operations. The raw config underneath is still accessible.

Closes #1309

### CODE REVIEW

File: common/es/src/main/java/io/apiman/common/es/util/DefaultEsClientFactory.java [hunk 1]
```diff
@@ -123,55 +124,51 @@ private RestHighLevelClient createEsClient(ApimanEsClientOptionsParser opts,
...
-                    clients.put(clientKey, client);
-                    LOGGER.info("Created new elasticsearch-client for {0}://{1}:{2} for index prefix {3}",
+                EsConnectionPoller esConnectionPoller = new EsConnectionPoller(
+                    client, 0, POLL_INTERVAL_SECS, Math.toIntExact(opts.getPollingTime())
+                );
+
+                // Block and wait for Elasticsearch. Exception will be raised if not successful.
+                esConnectionPoller.blockUntilReady();
+
+                // Put client to list if polling is successful
```
Comment (msavy) [inline common/es/src/main/java/io/apiman/common/es/util/DefaultEsClientFactory.java:134]: into

**Review by msavy — COMMENTED** [review]
self-review OK

**Review by volkflo — COMMENTED** [review]
I think this looks good :) I did not test it yet.

But I have an additional idea about the aliases:

I think it would be helpful to print a warning if the user uses an alias.
Like 
```java
keys("key1", "key2")

if (we match key2)
	Print: "You are not using an aliased config value. Support for this alias could be dropped during releases, please change the config to <key1>
```
This makes it easier for us to drop support for old/unused config and helps the user "updating" the config files over time.

What do you think?

File: common/plugin/src/test/java/io/apiman/common/config/options/GenericOptionsParserTest.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Test;
+import static org.assertj.core.api.Assertions.*;
+
+public class GenericOptionsParserTest {
+
+    @Test
+    public void Given_OptionMap_When_GettingGettingKey_Then_ShouldFindValueCaseInsensitively() {
```
[the remark this answers is not in the document]
Comment (msavy) [reply common/plugin/src/test/java/io/apiman/common/config/options/GenericOptionsParserTest.java:32]: lol

File: common/config/src/main/java/io/apiman/common/config/options/exceptions/BadOptionConfigurationException.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+     * Indicate a parsing failure.
+     *
+     * @param optionName   option name/key being parsed.
+     * @param expectedType the anticipated type (e.g. boolean).
+     * @param actualValue  the actual found.
+     * @param cause        the cause of this issue, if it is an exception (e.g. NumberFormatException).
+     */
+    public static BadOptionConfigurationException parseFailure(String optionName, String expectedType,
+        String actualValue, Throwable cause) {
+
```
[the remark this answers is not in the document]
Comment (msavy) [reply common/config/src/main/java/io/apiman/common/config/options/exceptions/BadOptionConfigurationException.java:54]: 👍💪

File: common/es/src/main/java/io/apiman/common/es/util/DefaultEsClientFactory.java [hunk 4]
```diff
@@ -69,126 +69,141 @@ public RestHighLevelClient createClient(Map<String, String> config,
...
-            String trustStorePath = config.get("client.truststore");
-            String trustStorePassword = config.get("client.truststore.password");
-
-            Path trustStorePathObject = Paths.get(trustStorePath);
-            Path keyStorePathObject = Paths.get(clientKeystorePath);
-            KeyStore truststore = KeyStore.getInstance("pkcs12");
-            KeyStore keyStore = KeyStore.getInstance("pkcs12");
-
-            try (InputStream is = Files.newInputStream(trustStorePathObject)) {
+            // TODO(msavy): merge together with TLSOptions?
```
[the remark this answers is not in the document]
Comment (msavy) [reply common/es/src/main/java/io/apiman/common/es/util/DefaultEsClientFactory.java:158]: https://github.com/apiman/apiman/issues/1314

File: common/plugin/src/test/java/io/apiman/common/config/options/GenericOptionsParserTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (msavy) [reply common/plugin/src/test/java/io/apiman/common/config/options/GenericOptionsParserTest.java:32]: I'll fix fix it 😂

**Comment by msavy** [comment]
> I think this looks good :) I did not test it yet.
> 
> But I have an additional idea about the aliases:
> 
> I think it would be helpful to print a warning if the user uses an alias.
> Like
> 
> ```java
> keys("key1", "key2")
> 
> if (we match key2)
> 	Print: "You are using an aliased config value. Support for this alias could be dropped during releases, please change the config to <key1>
> ```
> 
> This makes it easier for us to drop support for old/unused config and helps the user "updating" the config files over time.
> 
> What do you think?

Let's have a think about that and make a new ticket perhaps.

I think there are currently 2 uses for it:
- Deprecating old keys
- Aliasing keys where there's a `.property` pattern that can't be replicated in JSON - that's currently what I'm using it for. We probably need to distinguish between the two in any error cases.

**Review by volkflo — APPROVED** [review]
